File: js/problem10.js
//solution: 142913828922
print(eulerProblem10());

function eulerProblem10() {
    var sum = 2;
    for(var i = 3; i < 2000000; ++i) {
        if(isPrime(i)) {
            sum += i;
        }
    }
    return sum;
}

function isPrime(n) {
    if(n < 99) {
        return slowPrime(n);
    }
    
    if(n % 2 == 0
    || n % 3 == 0
    || n % 5 == 0
    || n % 7 == 0
    || n % 11 == 0
    || n % 13 == 0
    || n % 17 == 0
    || n % 19 == 0
    || n % 23 == 0
    || n % 29 == 0
    || n % 31 == 0
    || n % 37 == 0
    || n % 41 == 0
    || n % 43 == 0
    || n % 47 == 0
    || n % 53 == 0
    || n % 59 == 0
    || n % 61 == 0
    || n % 67 == 0
    || n % 71 == 0
    || n % 73 == 0
    || n % 79 == 0
    || n % 83 == 0
    || n % 89 == 0
    || n % 97 == 0) {
        return false;
    }
    
    for(var i = 3; i <= Math.sqrt(n); i += 2) {
        if(n % i == 0) {
            return false;
        }
    }
    
    return true;
}

function slowPrime(n) {
    if(n == 0) { return false; }
    if(n == 1) { return false; }
    if(n == 2) { return true; }
    if(n % 2 == 0) { return false; }
    for(var i = 2; i <= Math.sqrt(n); ++i) {
        if(n % i == 0) {
            return false;
        }
    }
    return true;
}

File: python/problem10.py
#!/usr/bin/python
#solution: 

import math

def is_prime(n):
    if n < 2:
        return False
    if n == 2:
        return True
    if n < 10:
        values = { \
            2: True, \
            3: True, \
            4: False, \
            5: True, \
            6: False, \
            7: True, \
            8: False, \
            9: False \
        }
        return values.get(n)
    i = 3
    while i <= math.sqrt(n):
        if n % i == 0:
            return False
        i += 2
    return True

def primes():
    p = 2
    while 1:
        if is_prime(p):
            yield p
        if p == 2:
            p = 3
        else:
            p += 2


def euler_problem_10():
    p = 0
    sum = 0
    pnums = primes()
    while p < 2000000:
        p = pnums.next()
        sum += p
    return sum - p

print euler_problem_10()
